### ISSUE TITLE
False positive with dereference.of.nullable

### ISSUE THREAD

**Issue by mahfouz72** [typetools/checker-framework#6760]
The Checker Framework is flagging a possible null dereference in the below code snippet, where `lambdaParameters.peek()` is being checked for null before it is dereferenced. This seems to be a false positive, as the code should not cause a null dereference.

Code:
 
```Java

while (lambdaParameters.peek() != null
        && ast.equals(lambdaParameters.peek().enclosingLambda)) {

    final Optional<LambdaParameterDetails> unusedLambdaParameter =
            Optional.ofNullable(lambdaParameters.peek())
                    .filter(parameter -> !parameter.isUsed())
                    .filter(parameter -> !"_".equals(parameter.getName()));

    ....
}
```
---

Output: 
```XML
New surviving error(s) found:

  <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java</fileName>
    <specifier>dereference.of.nullable</specifier>
    <message>dereference of possibly-null reference lambdaParameters.peek()</message>
    <lineContent>&amp;&amp; ast.equals(lambdaParameters.peek().enclosingLambda)) {</lineContent>
  </checkerFrameworkError>
```

---

Expectations:

This issue appears to be a false positive because` lambdaParameters.peek()` is explicitly checked for nullity before any dereferencing occurs. The error persists even though the code logic should prevent a null dereference.

---
Build tool: maven
checker version: 3.46.0

**Comment by smillst** [comment]
You've not given me enough of the code to reproduce this error.  It could be that `peek` is not marked ` @SideEffectFree` so `lambdaParameters.peek()` is unrefined.

**Comment by mahfouz72** [comment]
You can find the file [here](https://github.com/checkstyle/checkstyle/blob/602de17dbe665a8aceb05fe83f7d7ba99395a618/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLambdaParameterShouldBeUnnamedCheck.java)

`peek()` is a method in the Deque class from Java collections

**Comment by smillst** [comment]
Thanks!  I can reproduce this:

```
package typearginfer;

import java.util.ArrayDeque;
import java.util.Deque;

public class Issue6760 {
  private final Deque<LambdaParameterDetails> lambdaParameters = new ArrayDeque<>();
  void method(DetailAST ast) {
    while (lambdaParameters.peek() != null
        && ast.equals(lambdaParameters.peek().enclosingLambda)) {
    }
  }
  static class LambdaParameterDetails {
    private  DetailAST enclosingLambda = new DetailAST();
  }
  static class DetailAST{}
}
```

Yes, It seems like `peek` needs to be marked sideeffect free in the annotated jdk.

**Comment by mernst** [comment]
Fixed by https://github.com/typetools/jdk/pull/219.
